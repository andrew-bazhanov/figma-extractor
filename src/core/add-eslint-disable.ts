const maxLineRuleComment = '/* eslint-disable max-lines */';

const namingConventionRuleComment = '/* eslint-disable @typescript-eslint/naming-convention */';

export const addEslintDisableAtTheTopOfText = (
  content: string,
  options: {
    disabledMaxLines: boolean;
    disabledTypescriptNamingConvention: boolean;
  },
) => {
  return `
    ${options.disabledMaxLines ? maxLineRuleComment : ''}
    ${options.disabledTypescriptNamingConvention ? namingConventionRuleComment : ''}
    ${content}
    `;
};
