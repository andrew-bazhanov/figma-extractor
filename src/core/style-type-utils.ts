import { StyleType } from 'figma-js';

// eslint-disable-next-line @typescript-eslint/naming-convention
type StyleTypePredicate = (item: { style_type: StyleType }) => boolean;

const getStyleTypePredicate = (styleType: StyleType): StyleTypePredicate => {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  return ({ style_type }) => style_type === styleType;
};

export const styleTypeUtils = {
  isFill: getStyleTypePredicate('FILL'),
  isEffect: getStyleTypePredicate('EFFECT'),
  isText: getStyleTypePredicate('TEXT'),
};
