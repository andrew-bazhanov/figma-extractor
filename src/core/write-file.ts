import fs from 'fs';
import path from 'path';

const fileLengthRuleDisableComment = '/* eslint-disable max-lines */';

const namingConventionRuleDisableComment =
  '/* eslint-disable @typescript-eslint/naming-convention */';

const defaultOptions = {
  useEslintDisabledRules: true,
};

// NOTE: temporarily changed writeFile to writeFileSync to make testing easier:
// it is unclear how to test callback version without introducing promises.
// But even with promises, the logic seems to "just write" contents without returning the promise itself,
// which is confusing a bit, because it is impossible to handle promise fulfillment.
export const writeFile = (
  template: string,
  filePath: string,
  options?: {
    useEslintDisabledRules?: boolean;
  },
) => {
  const dir = path.dirname(filePath);
  const useEslintDisabledRules = { ...defaultOptions, ...options }.useEslintDisabledRules;

  // Create directory in advance if there is no one
  if (dir && !fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }

  const content = useEslintDisabledRules
    ? `
    ${fileLengthRuleDisableComment}
    ${namingConventionRuleDisableComment}
    ${template}
    `
    : template;

  try {
    fs.writeFileSync(filePath, content);
    console.log(`Generate ${filePath} is completed`);
  } catch {
    console.log(`Formatting ${filePath} file...`);
  }
};
