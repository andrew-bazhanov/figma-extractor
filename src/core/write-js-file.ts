import fs from 'fs';
import path from 'path';

const fileLengthRuleDisableComment = '/* eslint-disable max-lines */';

const namingConventionRuleDisableComment =
  '/* eslint-disable @typescript-eslint/naming-convention */';

// NOTE: temporarily changed writeFile to writeFileSync to make testing easier:
// it is unclear how to test callback version without introducing promises.
// But even with promises, the logic seems to "just write" contents without returning the promise itself,
// which is confusing a bit, because it is impossible to handle promise fulfillment.
export const writeJsFile = (template: string, filePath: string) => {
  const dir = path.dirname(filePath);

  // Create directory in advance if there is no one
  if (dir && !fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }

  try {
    fs.writeFileSync(
      filePath,
      `
      ${fileLengthRuleDisableComment}
      ${namingConventionRuleDisableComment}
      ${template}
      `,
    );
    console.log(`Generate ${filePath} is completed`);
  } catch {
    console.log(`Formatting ${filePath} file...`);
  }
};
