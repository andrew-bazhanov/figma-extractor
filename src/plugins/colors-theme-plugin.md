# colors-theme-plugin.ts

## How it works. Briefly.

Here it shows a few examples of how it works:

Before examples, we take such colors from API data:
```json
{
  "text/txt900":"#2a3fff",
  "light/text/txt900":"#ffffff",
  "dark/text/txt900":"#000000",
  "dark/text/txt700":"#00ff00",
  "monochrome/text/txt900":"#ff00ff",
  "colorful/text/txt900":"#00ffff",
  "text/txt800":"#3fc3fc3fc"
}
```
Plugin creates themes' files for only defined themes from a config, other styles will be ignored.

---

1. There are some themes and the default theme. 

Config:
```json
{
  "styles": {
    "colors": {
      "useTheme": true,
      "allowedThemes": ["dark","light","monochrome"],
      "defaultTheme": "monochrome"
    }
  }
}
```
It will generate such files:

* `themes-list.ts`:
```ts
export type DefaultTheme = 'monochrome';

export type Theme = 'light' | 'dark' | 'monochrome'; 
```
* `colors/dark/index.js`:
```js
module.exports = {
    "text-txt900":"#000000",
    "text-txt700":"#00ff00"
};
```
* `colors/light/index.js`:
```js
module.exports = {
    "text-txt900":"#ffffff",
};
```
* `colors/monochrome/index.js`:
```js
module.exports = {
    "text-txt900":"#ff00ff",
};
```
* `colors/index.js`:
```js
module.exports = {
    "text-txt900":"#ff00ff",
    "text-txt700":""
};
```
* `colors/with-vars.js`:
```js
module.exports = {
    "text-txt900":"var(--sh-text-txt900,'#ff00ff')",
    "text-txt700":"var(--sh-text-txt700,'')"
};
```
* `colors/dark/vars.css`:
```css
[data-theme='dark'] {
    --sh-text-txt900: '#000000';
    --sh-text-txt700: '#00ff00';
}
```
* `colors/light/vars.css`:
```css
[data-theme='light'] {
    --sh-text-txt900: '#ffffff';
    --sh-text-txt700: '';
}
```
* `colors/monochrome/vars.css`:
```css
[data-theme='monochrome'] {
    --sh-text-txt900: '#ff00ff';
    --sh-text-txt700: '';
}
```
* `colors/vars.css` (from *monochrome* theme):
```css
:root {
    --sh-text-txt900: '#ff00ff';
    --sh-text-txt700: '';
}
```
---
3. If there are some themes and one of them does not exist in API

Config:
```json
{
  "styles": {
    "colors": {
      "useTheme": true,
      "allowedThemes": ["dark-blue","light","monochrome"],
      "defaultTheme": "monochrome"
    }
  }
}
```
It will generate such files:

* `themes-list.ts`:
```ts
export type DefaultTheme = 'monochrome';

export type Theme = 'light' | 'dark-blue' | 'monochrome'; 
```
* `colors/dark-blue/index.js`:
```js
module.exports = {};
```
* `colors/light/index.js`:
```js
module.exports = {
    "text-txt900":"#ffffff",
};
```
* `colors/monochrome/index.js`:
```js
module.exports = {
    "text-txt900":"#ff00ff",
};
```
* `colors/index.js`:
```js
module.exports = {
    "text-txt900":"#ff00ff",
};
```
* `colors/with-vars.js`:
```js
module.exports = {
    "text-txt900":"var(--sh-text-txt900,'#ff00ff')",
};
```
* `colors/dark-blue/vars.css`:
```css
[data-theme='dark-blue'] {
    --sh-text-txt900: '';
}
```
* `colors/light/vars.css`:
```css
[data-theme='light'] {
    --sh-text-txt900: '#ffffff';
}
```
* `colors/monochrome/vars.css`:
```css
[data-theme='monochrome'] {
    --sh-text-txt900: '#ff00ff';
}
```
* `colors/vars.css` (from *monochrome* theme):
```css
:root {
    --sh-text-txt900: '#ff00ff';
}
```
