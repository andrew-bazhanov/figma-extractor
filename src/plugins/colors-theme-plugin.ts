import path from 'path';

import { getColorStyles } from '../lib/get-color-styles';
import { stringifyRecordsWithSort } from '../lib/stringify';
import { Plugin } from './types';
import { getColorName, replaceSlashToDash } from '../lib/color/get-color-name/get-color-name';

type ThemeName = string;

type ColorName = string;

type ColorValue = string;

type VariablesCollection = Record<ColorName, ColorValue>;

type ThemeCollection = Record<ThemeName, VariablesCollection>;

const generateJsVariables = (
  variablesCollection: VariablesCollection,
  defaultVariablesCollection: VariablesCollection,
) => {
  const data: Record<ColorName, string> = {};
  for (const colorName of Object.keys(variablesCollection)) {
    data[colorName] = `var(--sh-${colorName},'${defaultVariablesCollection[colorName] ?? ''}')`;
  }

  return data;
};

const generateCSSVariables = (variablesCollection: VariablesCollection, themeName?: ThemeName) => {
  const data: string[] = [];
  for (const [colorName, value] of Object.entries(variablesCollection)) {
    data.push(`--sh-${colorName}: '${value}';`);
  }
  data.sort((a, b) => {
    if (a === b) {
      return 0;
    }

    return a > b ? 1 : -1;
  });

  if (themeName && themeName !== '_') {
    return `
[data-theme='${themeName}'] {
  ${data.join('\n')}
}`;
  } else {
    return `
:root {
  ${data.join('\n')}
}`;
  }
};

const generateThemeListTS = (themeCollection: ThemeCollection, defaultTheme?: string) => {
  const themes: string[] = Object.keys(themeCollection)
    .filter(theme => {
      if (defaultTheme) {
        return theme !== defaultTheme && theme !== '_';
      }

      return theme !== '_';
    })
    .map(theme => `'${theme}'`);

  return `
// THIS FILE IS GENERATED AUTOMATICALLY. DON'T CHANGE IT.

export type DefaultTheme = '${defaultTheme ?? ''}';

export type Theme = ${themes.join(' | ')};  
  `;
};

export const colorsThemePlugin: Plugin = (
  { config, styleTypeUtils, writeFile, runFormattingFile },
  { styleMetadata, fileNodes },
) => {
  const metaColors = styleMetadata.filter(styleTypeUtils.isFill);

  if (config?.styles?.colors?.disabled) {
    return;
  }

  const allowedThemes = config?.styles?.colors?.allowedThemes;
  const defaultTheme = config?.styles?.colors?.defaultTheme;

  if (!Array.isArray(allowedThemes)) {
    throw new Error(
      '`config -> styles -> colors -> allowedThemes` field is required when the useTheme is equal true',
    );
  }

  if (allowedThemes.length === 0) {
    throw new Error(
      '`config -> styles -> colors -> allowedThemes` field must have one or more theme name',
    );
  }

  if (defaultTheme && !allowedThemes.includes(defaultTheme)) {
    throw new Error(
      "`config -> styles -> colors -> defaultTheme` field must be one of allowedThemes' values",
    );
  }

  const themesCollection = [...allowedThemes, '_'].reduce<ThemeCollection>((col, current) => {
    col[current] = {};

    return col;
  }, {});

  const colors = getColorStyles(metaColors, fileNodes, {
    ...config,
    styles: {
      ...config.styles,
      colors: {
        ...config.styles.colors,
        disabled: false,
        keyName: getColorName,
      },
    },
  });

  for (const [colorName, value] of Object.entries(colors)) {
    const separatedData = colorName.split('/');
    if (separatedData.length > 1) {
      const [separatedTheme, ...others] = separatedData;
      if (separatedTheme && allowedThemes.includes(separatedTheme)) {
        const newColorName =
          config?.styles?.colors?.keyName?.(others.join('/')) ??
          replaceSlashToDash(others.join('/'));

        themesCollection[separatedTheme][newColorName] = value;
        continue;
      }
    }

    const newColorName =
      config?.styles?.colors?.keyName?.(colorName) ?? replaceSlashToDash(colorName);

    themesCollection['_'][newColorName] = value;
  }

  for (const [themeName, variables] of Object.entries(themesCollection)) {
    const jsData = generateJsVariables(
      variables,
      defaultTheme && themesCollection[defaultTheme]
        ? themesCollection[defaultTheme]
        : themesCollection['_'],
    );

    const cssData = generateCSSVariables(variables, themeName);

    let fullPath = path.join(config?.styles?.exportPath || '', `colors/${themeName}`);
    if (defaultTheme && themesCollection[defaultTheme]) {
      if (themeName === '_') {
        continue;
      }
      fullPath = path.join(config?.styles?.exportPath || '', `colors`);
    } else if (themeName === '_') {
      fullPath = path.join(config?.styles?.exportPath || '', `colors`);
    }

    const jsTemplate = `module.exports = ${stringifyRecordsWithSort(jsData)};`;
    writeFile(jsTemplate, path.join(fullPath, 'index.js'));
    runFormattingFile(path.join(fullPath, 'index.js'));

    writeFile(cssData, path.join(fullPath, 'vars.css'));
    runFormattingFile(path.join(fullPath, 'vars.css'));
  }

  const themeListTSData = generateThemeListTS(themesCollection, defaultTheme);

  writeFile(themeListTSData, path.join(config?.styles?.exportPath || '', 'themes-list.ts'));
  runFormattingFile(path.join(config?.styles?.exportPath || '', `themes-list.ts`));
};

colorsThemePlugin.pluginName = 'colors-theme';
