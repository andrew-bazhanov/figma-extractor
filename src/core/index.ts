import { ClientInterface, FileNodesResponse, FullStyleMetadata } from 'figma-js';

import { Plugin } from '../plugins';
import { runFormattingFile } from './run-formatting-file';
import { styleTypeUtils } from './style-type-utils';
import { writeJsFile } from './write-js-file';

export interface FigmaData {
  figmaClient: ClientInterface;
  styleMetadata: readonly FullStyleMetadata[];
  fileNodes: FileNodesResponse;
}

export interface Core {
  config: Config;
  rootPath: string;
  plugins: Plugin[];
  styleTypeUtils: typeof styleTypeUtils;
  writeJsFile: typeof writeJsFile;
  runFormattingFile: typeof runFormattingFile;
}

export const createCore = (args: { config: Config; rootPath: string; plugins: Plugin[] }): Core => {
  return {
    ...args,
    styleTypeUtils,
    writeJsFile,
    runFormattingFile: runFormattingFile,
  };
};
