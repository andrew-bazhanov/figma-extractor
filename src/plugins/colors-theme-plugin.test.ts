/* eslint-disable max-lines */
import { createCore } from '../core';
import { fileNodes } from '../utils/generate-styles/fixtures/file-nodes';
import { styleMetadata } from '../utils/generate-styles/fixtures/style-metadata';
import { colorsThemePlugin } from './colors-theme-plugin';

describe('colorsThemePlugin', () => {
  it('should throw the error because the allowedThemes is not defined', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          colors: {
            useTheme: true,
          },
        },
      },
      plugins: [],
      rootPath: '/root-path',
      log: jest.fn(),
    });

    core.writeFile = jest.fn();

    expect(() =>
      colorsThemePlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any),
    ).toThrow(
      '`config -> styles -> allowedThemes` field is required when the useTheme is equal true',
    );
  });

  it('should throw the error because the defaultTheme is not defined', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          allowedThemes: [],
          colors: {
            useTheme: true,
          },
        },
      },
      plugins: [],
      rootPath: '/root-path',
      log: jest.fn(),
    });

    core.writeFile = jest.fn();

    expect(() =>
      colorsThemePlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any),
    ).toThrow(
      '`config -> styles -> defaultTheme` field is required when the useTheme is equal true',
    );
  });

  it('should throw the error because the allowedThemes is empty', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          allowedThemes: [],
          defaultTheme: 'any',
          colors: {
            useTheme: true,
          },
        },
      },
      plugins: [],
      rootPath: '/root-path',
      log: jest.fn(),
    });

    core.writeFile = jest.fn();

    expect(() =>
      colorsThemePlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any),
    ).toThrow('`config -> styles -> allowedThemes` field must have one or more theme name');
  });

  it('should throw the error because the defined defaultTheme is not included in allowedThemes list', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          allowedThemes: ['light', 'dark', 'monochrome'],
          defaultTheme: 'another',
          colors: {
            useTheme: true,
          },
        },
      },
      plugins: [],
      rootPath: '/root-path',
      log: jest.fn(),
    });

    core.writeFile = jest.fn();

    expect(() =>
      colorsThemePlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any),
    ).toThrow("`config -> styles -> defaultTheme` field must be one of allowedThemes' values");
  });

  it("should throw the error because any theme wasn't found inside figma data", () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          allowedThemes: ['theme1', 'theme2'],
          defaultTheme: 'theme1',
          colors: {
            useTheme: true,
          },
        },
      },
      plugins: [],
      rootPath: '/root-path',
      log: jest.fn(),
    });

    core.writeFile = jest.fn();

    expect(() =>
      colorsThemePlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any),
    ).toThrow("None of themes was found inside figma data. Check your themes and figma's themes");
  });

  it('should throw the error because some colors from themes contain not-valid chars', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          allowedThemes: ['light', 'dark', 'monochrome'],
          defaultTheme: 'monochrome',
          colors: {
            useTheme: true,
            keyName: (name?: string) => (name === 'light/text/txt900' ? name + '+prefix' : name),
          },
        },
      },
      plugins: [],
      rootPath: '/root-path',
      log: jest.fn(),
    });

    core.writeFile = jest.fn();

    expect(() =>
      colorsThemePlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any),
    ).toThrow('Color name: "text-txt900+prefix" from "light" theme contains not-valid chars.');
  });

  it('should throw the error because some colors from no themes contain not-valid chars', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          allowedThemes: ['light', 'dark', 'monochrome'],
          defaultTheme: 'monochrome',
          colors: {
            useTheme: true,
            keyName: (name?: string) => (name === 'text/txt900' ? name + '+prefix' : name),
          },
        },
      },
      plugins: [],
      rootPath: '/root-path',
      log: jest.fn(),
    });

    core.writeFile = jest.fn();

    expect(() =>
      colorsThemePlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any),
    ).toThrow('Color name: "text-txt900+prefix" without theme contains not-valid chars.');
  });

  it('should check the second parameter of the custom keyName function', () => {
    const keyName = jest.fn((name?: string) => name);
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          allowedThemes: ['light', 'dark', 'monochrome'],
          defaultTheme: 'monochrome',
          colors: {
            useTheme: true,
            keyName,
          },
        },
      },
      plugins: [],
      rootPath: '/root-path',
      log: jest.fn(),
    });

    core.writeFile = jest.fn();

    colorsThemePlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any);

    expect((keyName as jest.Mock).mock.calls[0][1]).toBeTruthy();
  });

  it('should create js files with css variables', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          allowedThemes: ['light', 'dark', 'monochrome'],
          defaultTheme: 'monochrome',
          colors: {
            useTheme: true,
          },
        },
      },
      plugins: [],
      rootPath: '/root-path',
      log: jest.fn(),
    });

    core.writeFile = jest.fn();

    colorsThemePlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any);

    expect(core.writeFile).toHaveBeenCalled();
    expect((core.writeFile as jest.Mock).mock.calls[0][1]).toBe(
      '/export-path/colors/light/index.js',
    );
    expect((core.writeFile as jest.Mock).mock.calls[0][0]).toMatchSnapshot(
      '/export-path/colors/light/index.js',
    );
    expect((core.writeFile as jest.Mock).mock.calls[2][1]).toBe(
      '/export-path/colors/dark/index.js',
    );
    expect((core.writeFile as jest.Mock).mock.calls[2][0]).toMatchSnapshot(
      '/export-path/colors/dark/index.js',
    );
    expect((core.writeFile as jest.Mock).mock.calls[4][1]).toBe(
      '/export-path/colors/monochrome/index.js',
    );
    expect((core.writeFile as jest.Mock).mock.calls[4][0]).toMatchSnapshot(
      '/export-path/colors/monochrome/index.js',
    );
    expect((core.writeFile as jest.Mock).mock.calls[6][1]).toBe('/export-path/colors/with-vars.js');
    expect((core.writeFile as jest.Mock).mock.calls[6][0]).toMatchSnapshot(
      '/export-path/colors/with-vars.js',
    );
    expect((core.writeFile as jest.Mock).mock.calls[7][1]).toBe('/export-path/colors/index.js');
    expect((core.writeFile as jest.Mock).mock.calls[7][0]).toMatchSnapshot(
      '/export-path/colors/index.js',
    );
  });

  it('should create css files with css variables', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          allowedThemes: ['light', 'dark', 'monochrome'],
          defaultTheme: 'monochrome',
          colors: {
            useTheme: true,
          },
        },
      },
      plugins: [],
      rootPath: '/root-path',
      log: jest.fn(),
    });

    core.writeFile = jest.fn();

    colorsThemePlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any);

    expect(core.writeFile).toHaveBeenCalled();
    expect((core.writeFile as jest.Mock).mock.calls[1][1]).toBe(
      '/export-path/colors/light/vars.css',
    );
    expect((core.writeFile as jest.Mock).mock.calls[1][0]).toMatchSnapshot(
      '/export-path/colors/light/vars.css',
    );
    expect((core.writeFile as jest.Mock).mock.calls[3][1]).toBe(
      '/export-path/colors/dark/vars.css',
    );
    expect((core.writeFile as jest.Mock).mock.calls[3][0]).toMatchSnapshot(
      '/export-path/colors/dark/vars.css',
    );
    expect((core.writeFile as jest.Mock).mock.calls[5][1]).toBe(
      '/export-path/colors/monochrome/vars.css',
    );
    expect((core.writeFile as jest.Mock).mock.calls[5][0]).toMatchSnapshot(
      '/export-path/colors/monochrome/vars.css',
    );
    expect((core.writeFile as jest.Mock).mock.calls[8][1]).toBe('/export-path/colors/vars.css');
    expect((core.writeFile as jest.Mock).mock.calls[8][0]).toMatchSnapshot(
      '/export-path/colors/vars.css',
    );
  });

  it("should create theme-list.ts files with themes' types", () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          allowedThemes: ['light', 'dark', 'monochrome'],
          defaultTheme: 'monochrome',
          colors: {
            useTheme: true,
          },
        },
      },
      plugins: [],
      rootPath: '/root-path',
      log: jest.fn(),
    });

    core.writeFile = jest.fn();

    colorsThemePlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any);

    expect(core.writeFile).toHaveBeenCalled();
    expect((core.writeFile as jest.Mock).mock.calls[9][1]).toBe('/export-path/themes-list.ts');
    expect((core.writeFile as jest.Mock).mock.calls[9][0]).toMatchSnapshot(
      '/export-path/themes-list.ts',
    );
  });

  describe("create theme data with one theme which doesn't exist in figma", () => {
    it('should create js files with css variables', () => {
      const core = createCore({
        config: {
          styles: {
            exportPath: '/export-path/',
            allowedThemes: ['light', 'dark-blue', 'monochrome'],
            defaultTheme: 'monochrome',
            colors: {
              useTheme: true,
            },
          },
        },
        plugins: [],
        rootPath: '/root-path',
        log: jest.fn(),
      });

      core.writeFile = jest.fn();

      colorsThemePlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any);

      expect(core.writeFile).toHaveBeenCalled();
      expect((core.writeFile as jest.Mock).mock.calls[0][1]).toBe(
        '/export-path/colors/light/index.js',
      );
      expect((core.writeFile as jest.Mock).mock.calls[0][0]).toMatchSnapshot(
        '/export-path/colors/light/index.js',
      );
      expect((core.writeFile as jest.Mock).mock.calls[2][1]).toBe(
        '/export-path/colors/dark-blue/index.js',
      );
      expect((core.writeFile as jest.Mock).mock.calls[2][0]).toMatchSnapshot(
        '/export-path/colors/dark-blue/index.js',
      );
      expect((core.writeFile as jest.Mock).mock.calls[4][1]).toBe(
        '/export-path/colors/monochrome/index.js',
      );
      expect((core.writeFile as jest.Mock).mock.calls[4][0]).toMatchSnapshot(
        '/export-path/colors/monochrome/index.js',
      );
      expect((core.writeFile as jest.Mock).mock.calls[6][1]).toBe(
        '/export-path/colors/with-vars.js',
      );
      expect((core.writeFile as jest.Mock).mock.calls[6][0]).toMatchSnapshot(
        '/export-path/colors/with-vars.js',
      );
      expect((core.writeFile as jest.Mock).mock.calls[7][1]).toBe('/export-path/colors/index.js');
      expect((core.writeFile as jest.Mock).mock.calls[7][0]).toMatchSnapshot(
        '/export-path/colors/index.js',
      );
    });

    it('should create css files with css variables', () => {
      const core = createCore({
        config: {
          styles: {
            exportPath: '/export-path/',
            allowedThemes: ['light', 'dark-blue', 'monochrome'],
            defaultTheme: 'monochrome',
            colors: {
              useTheme: true,
            },
          },
        },
        plugins: [],
        rootPath: '/root-path',
        log: jest.fn(),
      });

      core.writeFile = jest.fn();

      colorsThemePlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any);

      expect(core.writeFile).toHaveBeenCalled();

      expect((core.writeFile as jest.Mock).mock.calls[1][1]).toBe(
        '/export-path/colors/light/vars.css',
      );
      expect((core.writeFile as jest.Mock).mock.calls[1][0]).toMatchSnapshot(
        '/export-path/colors/light/vars.css',
      );
      expect((core.writeFile as jest.Mock).mock.calls[3][1]).toBe(
        '/export-path/colors/dark-blue/vars.css',
      );
      expect((core.writeFile as jest.Mock).mock.calls[3][0]).toMatchSnapshot(
        '/export-path/colors/dark-blue/vars.css',
      );
      expect((core.writeFile as jest.Mock).mock.calls[5][1]).toBe(
        '/export-path/colors/monochrome/vars.css',
      );
      expect((core.writeFile as jest.Mock).mock.calls[5][0]).toMatchSnapshot(
        '/export-path/colors/monochrome/vars.css',
      );
      expect((core.writeFile as jest.Mock).mock.calls[8][1]).toBe('/export-path/colors/vars.css');
      expect((core.writeFile as jest.Mock).mock.calls[8][0]).toMatchSnapshot(
        '/export-path/colors/vars.css',
      );
    });

    it("should create theme-list.ts files with themes' types", () => {
      const core = createCore({
        config: {
          styles: {
            exportPath: '/export-path/',
            allowedThemes: ['light', 'dark-blue', 'monochrome'],
            defaultTheme: 'monochrome',
            colors: {
              useTheme: true,
            },
          },
        },
        plugins: [],
        rootPath: '/root-path',
        log: jest.fn(),
      });

      core.writeFile = jest.fn();

      colorsThemePlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any);

      expect(core.writeFile).toHaveBeenCalled();
      expect((core.writeFile as jest.Mock).mock.calls[9][1]).toBe('/export-path/themes-list.ts');
      expect((core.writeFile as jest.Mock).mock.calls[9][0]).toMatchSnapshot(
        '/export-path/themes-list.ts',
      );
    });
  });
});
