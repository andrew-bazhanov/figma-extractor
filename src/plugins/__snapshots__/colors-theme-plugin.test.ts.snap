// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`colorsThemePlugin create theme data with one theme which doesn't exist in figma should create css files with css variables: /export-path/colors/dark-blue/vars.css 1`] = `
"
[data-theme='dark-blue'] {
  --sh-text-txt900: '';
}"
`;

exports[`colorsThemePlugin create theme data with one theme which doesn't exist in figma should create css files with css variables: /export-path/colors/light/vars.css 1`] = `
"
[data-theme='light'] {
  --sh-text-txt900: '#ffffff';
}"
`;

exports[`colorsThemePlugin create theme data with one theme which doesn't exist in figma should create css files with css variables: /export-path/colors/monochrome/vars.css 1`] = `
"
[data-theme='monochrome'] {
  --sh-text-txt900: '#ff00ff';
}"
`;

exports[`colorsThemePlugin create theme data with one theme which doesn't exist in figma should create css files with css variables: /export-path/colors/vars.css 1`] = `
"
:root {
  --sh-text-txt900: '#ff00ff';
}"
`;

exports[`colorsThemePlugin create theme data with one theme which doesn't exist in figma should create js files with css variables: /export-path/colors/dark-blue/index.js 1`] = `
"
    /* eslint-disable max-lines */
    
    module.exports = {};
    "
`;

exports[`colorsThemePlugin create theme data with one theme which doesn't exist in figma should create js files with css variables: /export-path/colors/index.js 1`] = `
"
    /* eslint-disable max-lines */
    
    module.exports = {"text-txt900":"var(--sh-text-txt900,'#ff00ff')"};
    "
`;

exports[`colorsThemePlugin create theme data with one theme which doesn't exist in figma should create js files with css variables: /export-path/colors/light/index.js 1`] = `
"
    /* eslint-disable max-lines */
    
    module.exports = {"text-txt900":"#ffffff"};
    "
`;

exports[`colorsThemePlugin create theme data with one theme which doesn't exist in figma should create js files with css variables: /export-path/colors/monochrome/index.js 1`] = `
"
    /* eslint-disable max-lines */
    
    module.exports = {"text-txt900":"#ff00ff"};
    "
`;

exports[`colorsThemePlugin create theme data with one theme which doesn't exist in figma should create theme-list.ts files with themes' types: /export-path/themes-list.ts 1`] = `
"
    /* eslint-disable max-lines */
    /* eslint-disable @typescript-eslint/naming-convention */
    
// THIS FILE IS GENERATED AUTOMATICALLY. DON'T CHANGE IT.

export type DefaultTheme = 'monochrome';

export type Theme = 'light' | 'dark-blue' | 'monochrome';  
  
    "
`;

exports[`colorsThemePlugin should create css files with css variables: /export-path/colors/dark/vars.css 1`] = `
"
[data-theme='dark'] {
  --sh-text-txt700: '#00ff00';
--sh-text-txt900: '#000000';
}"
`;

exports[`colorsThemePlugin should create css files with css variables: /export-path/colors/light/vars.css 1`] = `
"
[data-theme='light'] {
  --sh-text-txt700: '';
--sh-text-txt900: '#ffffff';
}"
`;

exports[`colorsThemePlugin should create css files with css variables: /export-path/colors/monochrome/vars.css 1`] = `
"
[data-theme='monochrome'] {
  --sh-text-txt700: '';
--sh-text-txt900: '#ff00ff';
}"
`;

exports[`colorsThemePlugin should create css files with css variables: /export-path/colors/vars.css 1`] = `
"
:root {
  --sh-text-txt700: '';
--sh-text-txt900: '#ff00ff';
}"
`;

exports[`colorsThemePlugin should create js files with css variables: /export-path/colors/dark/index.js 1`] = `
"
    /* eslint-disable max-lines */
    
    module.exports = {"text-txt700":"#00ff00","text-txt900":"#000000"};
    "
`;

exports[`colorsThemePlugin should create js files with css variables: /export-path/colors/index.js 1`] = `
"
    /* eslint-disable max-lines */
    
    module.exports = {"text-txt700":"var(--sh-text-txt700,'')","text-txt900":"var(--sh-text-txt900,'#ff00ff')"};
    "
`;

exports[`colorsThemePlugin should create js files with css variables: /export-path/colors/light/index.js 1`] = `
"
    /* eslint-disable max-lines */
    
    module.exports = {"text-txt900":"#ffffff"};
    "
`;

exports[`colorsThemePlugin should create js files with css variables: /export-path/colors/monochrome/index.js 1`] = `
"
    /* eslint-disable max-lines */
    
    module.exports = {"text-txt900":"#ff00ff"};
    "
`;

exports[`colorsThemePlugin should create theme-list.ts files with themes' types: /export-path/themes-list.ts 1`] = `
"
    /* eslint-disable max-lines */
    /* eslint-disable @typescript-eslint/naming-convention */
    
// THIS FILE IS GENERATED AUTOMATICALLY. DON'T CHANGE IT.

export type DefaultTheme = 'monochrome';

export type Theme = 'light' | 'dark' | 'monochrome';  
  
    "
`;
